crates/ruma-events-macros/src/event_type.rs

```diff
@@ -0,0 +1,180 @@
+    let message = generate_enum(
+        format_ident!("MessageEventType"),
+        message_str_ev_types,
+        message_event_types,
+        &ruma_serde,
+    );
+
+    let (ephemeral_event_types, ephemeral_str_ev_types) = generate_variants(&ephemeral)?;
+    let ephemeral = generate_enum(
+        format_ident!("EphemeralRoomEventType"),
+        ephemeral_str_ev_types,
+        ephemeral_event_types,
+        &ruma_serde,
+    );
+
+    let (room_account_event_types, room_account_str_ev_types) = generate_variants(&room_account)?;
+    let room_account = generate_enum(
+        format_ident!("RoomAccountDataEventType"),
+        room_account_str_ev_types,
+        room_account_event_types,
+        &ruma_serde,
+    );
+
+    let (global_account_event_types, global_account_str_ev_types) =
+        generate_variants(&global_account)?;
+    let global_account = generate_enum(
+        format_ident!("GlobalAccountDataEventType"),
+        global_account_str_ev_types,
+        global_account_event_types,
+        &ruma_serde,
+    );
+
+    let (to_device_event_types, to_device_str_ev_types) = generate_variants(&to_device)?;
+    let to_device = generate_enum(
+        format_ident!("ToDeviceEventType"),
+        to_device_str_ev_types,
+        to_device_event_types,
+        &ruma_serde,
+    );
+
+    Ok(quote! {
+        #all
+
+        #room
+
+        #state
+
+        #message
+
+        #ephemeral
+
+        #room_account
+
+        #global_account
+
+        #to_device
+    })
+}
+
+fn generate_enum(
+    ident: Ident,
+    ev_type_strings: Vec<&LitStr>,
+    variants: Vec<TokenStream>,
+    ruma_serde: &TokenStream,
+) -> TokenStream {
+    let str_doc = format!("Creates a str slice from this `{}`.", ident);
+    let byte_doc = format!("Creates a byte slice from this `{}`.", ident);
+    let enum_doc = format!("The type of `{}` this is.", ident);
+    quote! {
+        #[doc = #enum_doc]
+        ///
+        /// This type can hold an arbitrary string. To check for events that are not available as a
+        /// documented variant here, use its string representation, obtained through `.as_str()`.
+        #[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, #ruma_serde::StringEnum)]
+        pub enum #ident {
+            #(
+                #[doc = #ev_type_strings]
+                #[ruma_enum(rename = #ev_type_strings)]
+                #variants,
+            )*
+            #[doc(hidden)]
+            _Custom(String),
+        }
+
+        impl #ident {
+            #[doc = #str_doc]
+            pub fn as_str(&self) -> &str {
+                self.as_ref()
+            }
+
+            #[doc = #byte_doc]
+            pub fn as_bytes(&self) -> &[u8] {
+                self.as_str().as_bytes()
+            }
+        }
+    }
+}
+
+fn generate_variants<'a>(
+    input: &'a [&Vec<EventEnumEntry>],
+) -> syn::Result<(Vec<TokenStream>, Vec<&'a LitStr>)> {
+    let mut deduped: Vec<&EventEnumEntry> = vec![];
```

Comment:
I don't understand what's going on here. Why does anything need to be deduplicated?

Reply:
`Message` and `ToDevice` have the same variants so when we combine them into `EventType` there are duplicate variants. I wasn't sure of the best way to go about doing this since `BTreeSet` and `HashSet` don't work great for the `EventEnumEntry` and the attributes makes it even more complicated.

So the idea is if we have duplicate variants keep the one without `cfg` attrs since we know that will for sure be there.

Reply:
Okay, if this is only used for `EventType` this sounds good 👍🏼

Reply:
Yeah, it's just an `EventType` thing.